### ISSUE TITLE
测试的时候，卡住不动了。

### ISSUE THREAD

**Issue by the-cat-crying** [xuelunshen/gim#18]
![image](https://github.com/user-attachments/assets/b72a12fb-4035-441f-9082-de252af83f63)

**Comment by xuelunshen** [comment]
有没有什么报错信息? 就只是卡住了吗?

**Comment by the-cat-crying** [comment]
就是没有报错误信息，就是卡住不动，只能强行KILL。好像是数据加载哪里的问题，单张测试没得问题。

**Comment by xuelunshen** [comment]
如果单卡测试没问题, 多卡测试有问题, 那可能和 ddp 有关. 因为 ZEB 测试的数据量有点大, 我用 Pytorch Lightning 写, 使其可以多机多卡测试.

**Comment by the-cat-crying** [comment]
可能没太表述清楚，就是使用你给的demo.py推理可以，就是使用“python test.py --gpus 1 --weight gim_lightlue --versiong 100h --test --batch_size 16 --tests        SceneNet”命令就不行了，卡住不动，也是使用的单卡，不知道是不是torch版本问题，
我的环境：
absl-py                 2.1.0
aiohttp                 3.9.5
aiosignal               1.3.1
albumentations          1.0.1
antlr4-python3-runtime  4.9.3
async-timeout           4.0.3
attrs                   23.2.0
certifi                 2024.7.4
charset-normalizer      3.3.2
contourpy               1.2.1
cycler                  0.12.1
einops                  0.3.0
fonttools               4.53.1
frozenlist              1.4.1
fsspec                  2024.6.1
future                  1.0.0
grpcio                  1.65.1
h5py                    3.11.0
idna                    3.7
imageio                 2.34.2
imagesize               1.2.0
joblib                  1.0.1
kiwisolver              1.4.5
kornia                  0.6.10
lazy_loader             0.4
lightning-utilities     0.11.6
loguru                  0.5.3
Markdown                3.6
MarkupSafe              2.1.5
matplotlib              3.7.5
multidict               6.0.5
networkx                3.3
numpy                   1.24.0
omegaconf               2.3.0
opencv-python-headless  4.10.0.84
packaging               24.1
pillow                  10.4.0
pip                     24.0
protobuf                4.25.4
pyDeprecate             0.3.1
pyparsing               3.1.2
python-dateutil         2.9.0.post0
pytorch-lightning       1.5.10
PyYAML                  6.0.1
requests                2.32.3
scikit-image            0.24.0
scipy                   1.14.0
setuptools              59.5.0
six                     1.16.0
tensorboard             2.17.0
tensorboard-data-server 0.7.2
tifffile                2024.7.24
torch                   1.11.0+cu113
torchaudio              0.11.0+cu113
torchmetrics            1.4.0.post0
torchvision             0.12.0+cu113
tqdm                    4.66.4
typing_extensions       4.12.2
urllib3                 2.2.2
Werkzeug                3.0.3
wheel                   0.43.0
yacs                    0.1.8
yarl                    1.9.4

**Comment by xuelunshen** [comment]
Torch 的版本要麻烦你去和我 README 里写的核对一下
因为 demo.py 是单纯的推理, 所以我没用 Pytorch Lightning
而 test.py 是用 Pytorch Lightning 写的
所以, 可能是 ddp 相关的设置问题
如果涉及版本兼容性的话, 需要注意的是 Pytorch Lightning 的版本必须是我指定的 1.5.3 这个版本
因为我一直是用的这个版本, 其他版本我不保证接口什么的发生了什么变化

**Comment by the-cat-crying** [comment]
我按照你的zeb步骤，下载图片压缩包，并解压到zeb中，还是没有成功，版本用的Pytorch Lightning==1.5.3，但是你README写的是1.5.10.
不知道是不是图片路径需要相应的更改
SceneNet 的数据集的cfg中参数
# DATA_ROOT = 'data/SceneNetRGBD/'
DATA_ROOT = '../gim-main/zeb/'
NPZ_ROOT = DATA_ROOT

_CN.NJOBS = 1

# TRAIN
_CN.DATASET.TRAIN = CN()
_CN.DATASET.TRAIN.PADDING = None
_CN.DATASET.TRAIN.DATA_ROOT = None
_CN.DATASET.TRAIN.NPZ_ROOT = None
_CN.DATASET.TRAIN.MAX_SAMPLES = None
_CN.DATASET.TRAIN.MIN_OVERLAP_SCORE = None
_CN.DATASET.TRAIN.MAX_OVERLAP_SCORE = None
_CN.DATASET.TRAIN.AUGMENTATION_TYPE = None
_CN.DATASET.TRAIN.LIST_PATH = None

# VALID
_CN.DATASET.VALID = CN()
_CN.DATASET.VALID.PADDING = None
_CN.DATASET.VALID.DATA_ROOT = None
_CN.DATASET.VALID.NPZ_ROOT = None
_CN.DATASET.VALID.MAX_SAMPLES = None
_CN.DATASET.VALID.MIN_OVERLAP_SCORE = None
_CN.DATASET.VALID.MAX_OVERLAP_SCORE = None
_CN.DATASET.VALID.AUGMENTATION_TYPE = None
_CN.DATASET.VALID.LIST_PATH = None

# TESTS
_CN.DATASET.TESTS = CN()
_CN.DATASET.TESTS.PADDING = False
_CN.DATASET.TESTS.DATA_ROOT = join(DATA_ROOT, 'test')
_CN.DATASET.TESTS.NPZ_ROOT = NPZ_ROOT
_CN.DATASET.TESTS.MAX_SAMPLES = 30
_CN.DATASET.TESTS.MIN_OVERLAP_SCORE = 0.0
_CN.DATASET.TESTS.MAX_OVERLAP_SCORE = 0.5
_CN.DATASET.TESTS.AUGMENTATION_TYPE = None
_CN.DATASET.TESTS.LIST_PATH = 'datasets/_tests_/SceneNetRGBD.txt'

cfg = _CN

这些地方，我发现datasets/_tests_/SceneNetRGBD.txt这txt文件在代码中并没有。

**Comment by the-cat-crying** [comment]
我把工作空间--threads设置为0就可以工作了

出现了这种Warning：/work/zhoup/conda_env/gim/lib/python3.10/site-packages/pytorch_lightning/trainer/data_loading.py:111: UserWarning: The dataloader, test_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 48 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.

看到一个类似的问题[https://github.com/IDEA-CCNL/Fengshenbang-LM/issues/123](url)
跟目前的问题很多相似，希望大佬可以帮我指点指点

**Comment by xuelunshen** [comment]
这个 UserWarning 你不用管它, 它是 Pytorch Lightning 提醒你说, num_workers (就是你刚刚设置的--threads) 你设为了 0, 如果这个参数设为 0, dataloader 就只会在主进程里面进行数据读取, 所以 Pytorch Lightning 提醒你这有可能会成为运行的速度瓶颈. 因为如果读取数据的速度慢了, 可能显卡要等数据读进显卡才能进行预测. 

如果你将 --threads 设为为 3 或者 8, 数据读取就会开 3 个或者 8 个子线程进行读取, 速度一般来说会比在主线程里读快一些.

不过, 你说只有 --threads 设置为 0 的时候才可以正常运行, 那你就设置为 0 吧, 慢点就慢点吧.

**Comment by the-cat-crying** [comment]
是的，大佬，能工作就是，所有我感觉还是dataloader哪里多线程还是有点问题，之前也遇到过，num_workers设置问题。

大佬可以用Hug个ingFace的Trainer的试一试不？我看这个[https://github.com/IDEA-CCNL/Fengshenbang-LM/issues/123](url)说是解决了。

**Comment by xuelunshen** [comment]
我这边跑的时候 num_workers 没啥问题的, 可能你还是环境和配置的关系. 因为我在很多台服务器重新配置过环境, 然后运行我根据 Pytorch Lightning 框架写的代码, 都没遇到 num_workers 出现啥问题.

**Comment by the-cat-crying** [comment]
确实，也有可能就是我的服务器环境问题，反正work就行。
感谢，gim_dkm的权重有百度云盘或者其他的么？google的上不去得要梯子。

**Comment by xuelunshen** [comment]
我传了一份到 [OneDrive](https://stuxmueducn-my.sharepoint.com/:u:/g/personal/xuelun_stu_xmu_edu_cn/EdJOibZ8VABOoKoyOHWo8ZEBHd_MyHbSvhRyT_o40SIPGA?e=GCjGZE), 这个国内应该可以直接访问.

**Comment by the-cat-crying** [comment]
好的，非常感谢大佬。
